src/analyzer/plugins/analyzerkeyfinder.cpp

```diff
@@ -0,0 +1,115 @@
+#include "analyzer/plugins/analyzerkeyfinder.h"
+
+#include "analyzer/constants.h"
+#include "util/assert.h"
+#include "util/math.h"
+
+using mixxx::track::io::key::ChromaticKey;
+using mixxx::track::io::key::ChromaticKey_IsValid;
+
+namespace {
+ChromaticKey chromaticKeyFromKeyFinderKeyT(KeyFinder::key_t key) {
+    switch (key) {
+    case (KeyFinder::A_MAJOR):
+        return ChromaticKey::A_MAJOR;
+    case (KeyFinder::A_MINOR):
+        return ChromaticKey::A_MINOR;
+    case (KeyFinder::B_FLAT_MAJOR):
+        return ChromaticKey::B_FLAT_MAJOR;
+    case (KeyFinder::B_FLAT_MINOR):
+        return ChromaticKey::B_FLAT_MINOR;
+    case (KeyFinder::B_MAJOR):
+        return ChromaticKey::B_MAJOR;
+    case (KeyFinder::B_MINOR):
+        return ChromaticKey::B_MINOR;
+    case (KeyFinder::C_MAJOR):
+        return ChromaticKey::C_MAJOR;
+    case (KeyFinder::C_MINOR):
+        return ChromaticKey::C_MINOR;
+    case (KeyFinder::D_FLAT_MAJOR):
+        return ChromaticKey::D_FLAT_MAJOR;
+    case (KeyFinder::D_FLAT_MINOR):
+        return ChromaticKey::C_SHARP_MINOR;
+    case (KeyFinder::D_MAJOR):
+        return ChromaticKey::D_MAJOR;
+    case (KeyFinder::D_MINOR):
+        return ChromaticKey::D_MINOR;
+    case (KeyFinder::E_FLAT_MAJOR):
+        return ChromaticKey::E_FLAT_MAJOR;
+    case (KeyFinder::E_FLAT_MINOR):
+        return ChromaticKey::E_FLAT_MINOR;
+    case (KeyFinder::E_MINOR):
+        return ChromaticKey::E_MINOR;
+    case (KeyFinder::F_MAJOR):
+        return ChromaticKey::F_MAJOR;
+    case (KeyFinder::F_MINOR):
+        return ChromaticKey::F_MINOR;
+    case (KeyFinder::G_FLAT_MAJOR):
+        return ChromaticKey::F_SHARP_MAJOR;
+    case (KeyFinder::G_FLAT_MINOR):
+        return ChromaticKey::F_SHARP_MINOR;
+    case (KeyFinder::G_MAJOR):
+        return ChromaticKey::G_MAJOR;
+    case (KeyFinder::G_MINOR):
+        return ChromaticKey::G_MINOR;
+    case (KeyFinder::A_FLAT_MAJOR):
+        return ChromaticKey::A_FLAT_MAJOR;
+    case (KeyFinder::A_FLAT_MINOR):
+        return ChromaticKey::G_SHARP_MINOR;
+    case (KeyFinder::SILENCE):
+    default:
+        return ChromaticKey::INVALID;
+    }
+}
+} // anonymous namespace
+
+namespace mixxx {
+
+AnalyzerKeyFinder::AnalyzerKeyFinder()
+        : m_keyFinder(),
+          m_workspace(),
+          m_audioData(),
+          m_currentFrame(0),
+          m_previousKey(mixxx::track::io::key::INVALID),
+          m_resultKeys() {
+}
+
+bool AnalyzerKeyFinder::initialize(int samplerate) {
+    m_audioData.setFrameRate(samplerate);
```

Comment:
What about the state and accumulated results when AnalyzerKeyFinder is reused for another analysis?

Reply:
If we need to reset this data we should use a single unique_ptr that holds an internal state struct with the non-resetable classes KeyFinder + Workspace + AudioData.

Reply:
I thought each analyzer instance was only used once?

Reply:
Yea, that's right.

Reply:
Okay so plain member variables initialized in the constructor should be fine.

Reply:
> What about the state and accumulated results when AnalyzerKeyFinder is reused for another analysis?

With such a concept all analyzer will benefit. We will receive a speed up and a improved accuracy. 
For instance changing key is not an exception. It happens almost always.
The key analysis picks a totally wrong key when passing a key change.
That's why it would be an improvement to detect beats and phrases fist.
Than use this info to detect the key in the phrase. This is usually constsnt.
From this information we can inturn improve the beat an phrase detection.
Normally the first chord in the new key is the down beat.

So you see everything is connected somhow. And only when all aspects ate discovered the analyser can decide right.

Reply:
I am really curious how much better the key analysis could be if we analyzed phrases rather than the whole track. I suspect the accuracy could go up significantly.